js/navigation/index.js

```diff
@@ -1,82 +1,62 @@
 // @flow
 
 import React, { Component } from 'react'
+import { BackHandler } from 'react-native'
 import { connect } from 'react-redux'
-import { StackNavigator, addNavigationHelpers } from 'react-navigation'
+import { addNavigationHelpers, NavigationActions } from 'react-navigation'
+import { createReduxBoundAddListener } from 'react-navigation-redux-helpers'
 
 import type { MapStateToProps } from 'react-redux'
-import type { NavigationState } from 'react-navigation/src/TypeDefinition'
+import type { NavigationState } from 'react-navigation'
 
-import ROUTES from './routes'
-import LoginScreen from '../components/LoginScreen'
-import ProfileScreen from '../components/ProfileScreen'
+import MainNavigator from './MainNavigator'
 
-import type { ApiUserProfile } from '../utils/api'
 import type { Dispatch } from '../actions/types'
 
-// Initialize the stack navigator
-const HomeRoutes = {
-  [ROUTES.HOME]: {
-    screen: LoginScreen
-  },
-
-  [ROUTES.PROFILE]: {
-    screen: ProfileScreen
-  }
-}
-
-export const HomeNavigator = StackNavigator(
-  {
-    ...HomeRoutes
-  },
-  {
-    initialRouteName: ROUTES.HOME,
-
-    // Let each screen handle the header and navigation
-    headerMode: 'none'
-  }
-)
-
-export const ProfileNavigator = StackNavigator(
-  {
-    ...HomeRoutes
-  },
-  {
-    initialRouteName: ROUTES.PROFILE,
-
-    // Let each screen handle the header and navigation
-    headerMode: 'none'
-  }
-)
+const addListener = createReduxBoundAddListener('root')
 
 type Props = {
-  profile: ApiUserProfile,
-  nav: NavigationState,
+  navigation: NavigationState,
   dispatch: Dispatch
 }
 
 /**
  * Main app navigator.
  */
 class Navigation extends Component<Props> {
-  render() {
-    const { profile } = this.props
-    const Navigator = profile ? ProfileNavigator : HomeNavigator
+  componentDidMount() {
+    BackHandler.addEventListener('hardwareBackPress', this.onBackPress)
+  }
+
+  componentWillUnmount() {
+    BackHandler.removeEventListener('hardwareBackPress', this.onBackPress)
+  }
+
+  onBackPress = () => {
+    const { dispatch, navigation } = this.props
+    if (navigation.index <= 1) {
+      BackHandler.exitApp()
+      return
```

Comment:
why here `undefined` is returned while on line 42 a boolean is returned? what's the return type of this function?
also add comments on what is going on here